src/neo/Network/P2P/TaskManager.cs

```diff
@@ -274,56 +225,68 @@ public static Props Props(NeoSystem system)
+            Header last = Blockchain.Singleton.HeaderCache.Last;
+            uint headerHeight = last?.Index ?? NativeContract.Ledger.CurrentIndex(snapshot);
+            UInt256 headerHash = last?.Hash ?? NativeContract.Ledger.CurrentHash(snapshot);
+            uint currentHeight = NativeContract.Ledger.CurrentIndex(snapshot);
+            // When the number of AvailableTasks is no more than 0, no pending tasks of InventoryType.Block, it should process pending the tasks of headers
+            // If not HeaderTask pending to be processed it should ask for more Blocks
+            if ((!HasHeaderTask || globalTasks[HeaderTaskHash] < MaxConncurrentTasks) && headerHeight < session.LastBlockIndex && !Blockchain.Singleton.HeaderCache.Full)
             {
-                if (lastTaskIndex >= highestBlockIndex || lastTaskIndex >= currentHeight + InvPayload.MaxHashesCount) break;
-                if (!AssignSyncTask(++lastTaskIndex)) break;
+                session.Tasks[HeaderTaskHash] = DateTime.UtcNow;
+                IncrementGlobalTask(HeaderTaskHash);
+                session.RemoteNode.Tell(Message.Create(MessageCommand.GetHeaders, GetBlockByIndexPayload.Create(headerHeight)));
             }
-        }
-
-        private void SendPingMessage()
-        {
-            DataCache snapshot = Blockchain.Singleton.View;
-            uint currentHeight = NativeContract.Ledger.CurrentIndex(snapshot);
-            UInt256 currentHash = NativeContract.Ledger.CurrentHash(snapshot);
-            TrimmedBlock block = NativeContract.Ledger.GetTrimmedBlock(snapshot, currentHash);
-            foreach (KeyValuePair<IActorRef, TaskSession> item in sessions)
+            else if (currentHeight < session.LastBlockIndex)
             {
-                var node = item.Key;
-                var session = item.Value;
-
-                if (session.ExpireTime < TimeProvider.Current.UtcNow ||
-                     (block.Index >= session.LastBlockIndex &&
-                     TimeProvider.Current.UtcNow.ToTimestampMS() - PingCoolingOffPeriod >= block.Timestamp))
+                UInt256 hash = NativeContract.Ledger.CurrentHash(snapshot);
+                for (uint i = currentHeight + 1; i <= headerHeight; i++)
                 {
-                    if (session.InvTasks.Remove(MemPoolTaskHash))
-                    {
-                        node.Tell(Message.Create(MessageCommand.Mempool));
-                    }
-                    node.Tell(Message.Create(MessageCommand.Ping, PingPayload.Create(currentHeight)));
-                    session.ExpireTime = TimeProvider.Current.UtcNow.AddMilliseconds(PingCoolingOffPeriod);
+                    UInt256 nextHash = Blockchain.Singleton.HeaderCache[i].Hash;
+                    if (!globalTasks.ContainsKey(nextHash)) break;
+                    hash = nextHash;
                 }
+                session.RemoteNode.Tell(Message.Create(MessageCommand.GetBlocks, GetBlocksPayload.Create(hash)));
```

Comment:
Why not use `GetBlockByIndex`?

Reply:
fixed

Reply:
No, I don't think so. You should use `GetBlockByIndex`, instead of changing `GetBlocks`.

Reply:
fixed